src/platform/datapath_winkernel.c

```diff
@@ -2034,8 +2101,34 @@ QuicDataPathSocketReceive(
                     goto Drop;
                 }
 
+                if (RecvContext->ProcContext->OutstandingStoredBytes > 256000) {
+                    //
+                    // Perform a copy
+                    //
+                    RecvContext->IsCopiedBuffer = TRUE;
+                    RecvContext->BufferPoolIndex = DataLength > 4096 ? 1 : 0;
+                    RecvContext->DataBufferStart =
+                        (uint8_t*)QuicPoolAlloc(
+                            &RecvContext->ProcContext->RecvBufferPools[RecvContext->BufferPoolIndex]);
+                    if (RecvContext->DataBufferStart == NULL) {
+                        QuicTraceLogWarning(
+                            DatapathDropAllocRecvBufferFailure,
+                            "[%p] Couldn't allocate receive buffers.",
+                            Binding);
+                        goto Drop;
+                    }
+                    CurrentCopiedBuffer = RecvContext->DataBufferStart;
+                } else {
+                    RecvContext->IsCopiedBuffer = FALSE;
+                    RecvContext->DataIndication = DataIndication;
+                    QUIC_DBG_ASSERT(DataIndication->Next == NULL);
+                    RecvContext->DataIndicationSize = (int32_t)DataLength;
+                    InterlockedAdd64(
+                        &RecvContext->ProcContext->OutstandingStoredBytes,
```

Comment:
I want to increment/decrement OutstandingStoredBytes regardless of if the bytes are pending or copied. I also want to be able to use this as an upper limit for number of bytes we're willing to pend before we start dropping.

Reply:
We have to have this variable be a pending buffer only list. Otherwise as soon as we cross the limit, it would basically be impossible to get out of, because if the pending buffers get released, that might not be enough to bring the stored bytes below the limit, so everything stays copy forever, even though we aren't actually pending anything.

I'll rename it to OutstandingPendingBytes, and if we want to add a separate OutstandingBytesInFlight we can easily enough.

Reply:
That assumes we don't drain any outstanding pending or copied data before receiving more. That's not a valid assumption necessarily.

Reply:
Well its not just draining any. To get below the threshold you'd have to drain the entire thing. Remember the pending buffers are going to be at the front of the queue, since they were received first. So they'll get returned first, while the copied ones are still pending. Then if any more packets are received before more are processed, they'll be copied instead of pended because we'll still be above the limit in flight, even though in flight is all copied packets.